src/wp-includes/class-wp-scripts.php

```diff
@@ -714,6 +777,185 @@ public function in_default_dir( $src ) {
+		foreach ( $this->registered as $registered_handle => $args ) {
+			if ( in_array( $handle, $args->deps, true ) ) {
+				$dependents[] = $registered_handle;
+			}
+		}
+		return $dependents;
+	}
+
+	/**
+	 * Get the strategy assigned during script registration.
+	 *
+	 * @param string $handle The script handle.
+	 * @return string|bool Strategy set during script registration. False if none was set.
+	 */
+	private function get_intended_strategy( $handle ) {
+		$script_args = $this->get_data( $handle, 'script_args' );
+		return isset( $script_args['strategy'] ) ? $script_args['strategy'] : false;
+	}
+
+	/**
+	 * Check if a script has a non standalone inline script associated with it.
+	 *
+	 * @param string $handle   The script handle.
+	 * @param string $position Position of the inline script.
+	 *
+	 * @return bool True if script present. False if empty.
+	 */
+	private function has_non_standalone_inline_script( $handle, $position ) {
+		$non_standalone_script_key = $position . '-non-standalone';
+		$non_standalone_script     = $this->get_data( $handle, $non_standalone_script_key );
+		return ! empty( $non_standalone_script );
+	}
+
+	/**
+	 * Check if all of a scripts dependents are deferrable, which is required to maintain execution order.
+	 *
+	 * @param string $handle  The script handle.
+	 * @param array $checked An array of already checked script handles, used to avoid looping recursion.
+	 * @return bool True if all dependents are deferrable, false otherwise.
+	 */
+	private function all_dependents_are_deferrable( $handle, $checked = array() ) {
+		// If this node was already checked, this script can be deferred and the branch ends.
+		if ( in_array( $handle, $checked, true ) ) {
+			return true;
+		}
+		$checked[]  = $handle;
+		$dependents = $this->get_dependents( $handle );
+
+		// If there are no dependents remaining to consider, the script can be deferred and the branch ends.
+		if ( empty( $dependents ) ) {
+			return true;
+		}
+
+		// Consider each dependent and check if it is deferrable.
+		foreach ( $dependents as $dependent ) {
+			// If the dependent script is not using the defer or async strategy, no script in the chain is deferrable.
+			if ( ! in_array( $this->get_intended_strategy( $dependent ), array( 'defer', 'async' ), true ) ) {
+				return false;
+			}
+
+			// If the dependent script has a non-standalone inline script in the 'before' position associated with it, do not defer.
+			if ( $this->has_non_standalone_inline_script( $dependent, 'before' ) ) {
+				return false;
+			}
+
+			// Recursively check all dependents.
+			if ( ! $this->all_dependents_are_deferrable( $dependent, $checked ) ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Get the most eligible loading strategy for a script.
+	 *
+	 * @param string  $handle The registered handle of the script.
+	 * @return string $strategy return the final strategy.
+	 */
+	private function get_eligible_loading_strategy( $handle = '' ) {
```

Comment:
Why is this an optional param? We should either remove that, or update the docblock accordingly.

```suggestion
	private function get_eligible_loading_strategy( $handle ) {
```

Reply:
Agreed, we won't be calling this without any handle. So removed the default.